---
title: "1.6 OG abundance importing"
output:
  html_document:
    fig_width: 6
    fig_height: 4
    df_print: paged
editor_options: 
  chunk_output_type: inline
---

# Importing RNA-seq reads

To capture the expression of in-paralogues, we import the count tables treating orthogroups (OGs) as genes and genes and isoforms as isoforms, according to the categorisation of `tximport`.

Thus, as you'd expect, we use the `tximport` library. We also use the script from [`OthoMantegazza/crossr`](https://github.com/othomantegazza/crossr/blob/master/R/parsers.R) called `parse_orthogroups.R`. This is needed because it was quicker to parse the `Orthogroup.tsv` file.

```{r}
library(tximport)
library(tidyverse)
library(myTAI)
source("functions/parse_orthogroups.R")
```

## Parse orthogroups

```{r}
OG_tx2gene <- 
  parse_orthogroups("data/Results_Nov29/Orthogroups/Orthogroups.tsv") %>% 
  stack() %>% 
  tidyr::as_tibble() %>% 
  dplyr::rename(GeneID = values) %>% 
  dplyr::rename(OG = ind) %>%
  dplyr::relocate(GeneID) %>%
  # dplyr::filter(str_detect(GeneID, "ser_|dis_")) %>%
  dplyr::mutate(GeneID = str_remove(GeneID, "\\.p[0-9].*"))
```

## Import OG as genes

### Abundance
```{r}
sample_info_fd <-
  readr::read_csv("data/sample_info_fd.csv")
sample_info_fs <-
  readr::read_csv("data/sample_info_fs.csv")
sample_info_ec <-
  readr::read_csv("data/sample_info_ec.csv")
```

```{r}
selectGenes <- function(counts, min.count=2){
  keep <-
    counts[rowMeans(counts)>min.count, ]
  return(keep)
}
```

```{r}
Ec_files <- 
  stringr::str_c(
    "data/Ec_salmon2/", 
    sample_info_ec$LibName, 
    "/quant.sf") %>%
  purrr::set_names(sample_info_ec$SampleName)

Ec_txi <- 
  tximport(
    Ec_files, 
    type = "salmon", 
    tx2gene = OG_tx2gene,
    countsFromAbundance = "lengthScaledTPM")
Ec_keep <- 
  selectGenes(Ec_txi$abundance, min.count = 2)
```

```{r}
Fd_files <- 
  stringr::str_c(
    "data/Fd_salmon/salmon/", 
    sample_info_fd$LibName, 
    "/quant.sf") %>%
  purrr::set_names(sample_info_fd$LibName)

# use abundance
Fd_txi <- 
  tximport(
    Fd_files, 
    type = "salmon", 
    tx2gene = OG_tx2gene,
    countsFromAbundance = "lengthScaledTPM")
Fd_keep <- 
  selectGenes(Fd_txi$abundance, min.count = 2)
```

```{r}
Fs_files <- 
  stringr::str_c(
    "data/Fs_salmon/salmon/", 
    sample_info_fs$LibName, 
    "/quant.sf") %>%
  purrr::set_names(sample_info_fs$LibName)

# use abundance
Fs_txi <- 
  tximport(
    Fs_files, 
    type = "salmon", 
    tx2gene = OG_tx2gene,
    countsFromAbundance = "lengthScaledTPM")
Fs_keep <- 
  selectGenes(Fs_txi$abundance, min.count = 2)
```

### Counts

```{r}
Fd_counts <-
  Fd_txi$counts[rownames(Fd_keep), ]
Fs_counts <-
  Fs_txi$counts[rownames(Fs_keep), ]
Ec_counts <-
  Ec_txi$counts[rownames(Ec_keep), ]
```

## Save OG abundance

```{r}
Fs_keep %>% 
  dplyr::as_tibble(rownames = "OG") %>%
  readr::write_csv(file = "data/Fs_OG_abundance.csv")

Fs_counts %>% 
  dplyr::as_tibble(rownames = "OG") %>%
  readr::write_csv(file = "data/Fs_OG_counts.csv")

Fd_keep %>% 
  dplyr::as_tibble(rownames = "OG") %>%
  readr::write_csv(file = "data/Fd_OG_abundance.csv")

Fd_counts %>% 
  dplyr::as_tibble(rownames = "OG") %>%
  readr::write_csv(file = "data/Fd_OG_counts.csv")

Ec_keep %>% 
  dplyr::as_tibble(rownames = "OG") %>%
  readr::write_csv(file = "data/Ec_OG_abundance.csv")

Ec_counts %>% 
  dplyr::as_tibble(rownames = "OG") %>%
  readr::write_csv(file = "data/Ec_OG_counts.csv")
```

# UpSet plot for OG size

Upset plot of orthogroups, as done in [Cossard et al. 2022 Nat Ecol Evol Extended Data Fig. 1](https://www.nature.com/articles/s41559-022-01692-4).

```{r}
# BiocManager::install("ComplexHeatmap")
library(ComplexHeatmap)
```

```{r}
assigned_OG <- parse_orthogroups("data/Results_Nov29/Orthogroups/Orthogroups.tsv") %>% stack() %>% tidyr::as_tibble() %>% 
  dplyr::rename(GeneID = values) %>% 
  dplyr::rename(OG = ind) %>%
  dplyr::relocate(GeneID) %>%
  dplyr::mutate(GeneID = str_remove(GeneID, "\\.[^.]+$")) %>%
  mutate(Species = case_when(
    str_detect(GeneID, "dis_") ~ "F. distichus",
    str_detect(GeneID, "ser_") ~ "F. serratus",
    str_detect(GeneID, "Ec-") ~ "Ectocarpus"
  )) %>%
  distinct() %>% # remove isoforms
  dplyr::select(OG, Species, GeneID)

unassigned_OG <- parse_orthogroups("data/Results_Nov29/Orthogroups/Orthogroups_UnassignedGenes.tsv") %>% stack() %>% tidyr::as_tibble() %>% 
  dplyr::rename(GeneID = values) %>% 
  dplyr::rename(OG = ind) %>%
  dplyr::relocate(GeneID) %>%
  dplyr::mutate(GeneID = str_remove(GeneID, "\\.[^.]+$")) %>%
  mutate(Species = case_when(
    str_detect(GeneID, "dis_") ~ "F. distichus",
    str_detect(GeneID, "ser_") ~ "F. serratus",
    str_detect(GeneID, "Ec-") ~ "Ectocarpus"
  )) %>%
  subset(!GeneID %in% assigned_OG$GeneID) %>%
  distinct(GeneID, .keep_all = TRUE) %>%
  dplyr::select(OG, Species, GeneID)

OG_species <- rbind(assigned_OG, unassigned_OG)
```

```{r}
list_names <- OG_species$Species
list_values <- OG_species$OG
list_OG_species <- unstack(OG_species, list_values ~ list_names)
```

```{r}
upsetting_input_distinct <- ComplexHeatmap::make_comb_mat(list_OG_species, mode = "distinct")
upsetting_input_intersect <- ComplexHeatmap::make_comb_mat(list_OG_species, mode = "intersect")
upsetting_input_union <- ComplexHeatmap::make_comb_mat(list_OG_species, mode = "union")


ComplexHeatmap::UpSet(upsetting_input_distinct,
                      row_title = "Species")
```

```{r}
upsetting_input_distinct
```

From `Statistics_Overall.tsv`

```bash
Number of species in orthogroup Number of orthogroups
1   2293
2   5590
3   2723
```

and from `Statistics_PerSpecies.tsv`

```bash
    F_distichus_pep_CA  F_serratus_pep_CA   PUBLIC_Ectocarpus-sp7_proteins
Number of genes 10213   13584   24428
Number of genes in orthogroups  7037    8630    16614
Number of unassigned genes  3176    4954    7814
Percentage of genes in orthogroups  68.9    63.5    68.0
Percentage of unassigned genes  31.1    36.5    32.0
Number of orthogroups containing species    6324    7031    8287
Percentage of orthogroups containing species    59.6    66.3    78.1
Number of species-specific orthogroups  62  181 2050
Number of genes in species-specific orthogroups 163 467 6630
Percentage of genes in species-specific orthogroups 1.6 3.4 27.1
```

Get session info.

```{r}
devtools::session_info()
```