---
title: "2.3 TAI enrichment"
output:
  html_document:
    fig_width: 6
    fig_height: 4
    df_print: paged
editor_options: 
  chunk_output_type: inline
---

# Are some PS enriched?

Using `myTAI::pStrata` -> `ggplot2::geom_col` or `ggplot2::geom_area` doesn't lead to interesting results. The way to do it is probably with a heatmap. `myTAI::pTAI` also doesn't look so nice.

```{r}
library(tidyverse)
library(myTAI)
#  myTAI v1.0.1.9000 2023-07-12 Github (drostlab/myTAI@46eb927)
```

# Playing around with `PhyloExpressionSetExample`

Based on this, I can intuit which type of analysis may be appropriate.

```{r}
data(PhyloExpressionSetExample)

heatmap(myTAI::pStrata(PhyloExpressionSetExample))

pStrata_data <-
  tibble::as_tibble(myTAI::pStrata(PhyloExpressionSetExample), rownames = "PS") %>%
  tidyr::pivot_longer(!PS, names_to = "Stage", values_to = "pStrata")

pStrata_data$Stage <-
  pStrata_data$Stage <- base::factor(pStrata_data$Stage, unique(pStrata_data$Stage))

pStrata_data %>%
  ggplot2::ggplot(aes(x = Stage, y = PS, fill = pStrata)) +
  ggplot2::geom_tile()
# 
# pStrata_data %>% 
#   ggplot2::ggplot(aes(x = Stage, y = pStrata, fill = PS)) +
#   ggplot2::geom_col()
# 
# pStrata_data2 <-
#   pStrata_data %>%
#   dplyr::mutate(
#     Stage_number = dplyr::case_when(
#       Stage == "Zygote" ~ 1,
#       Stage == "Quadrant" ~ 2,
#       Stage == "Globular" ~ 3,
#       Stage == "Heart" ~ 4,
#       Stage == "Torpedo" ~ 5,
#       Stage == "Bent" ~ 6,
#       Stage == "Mature" ~ 7
#     )
#   ) %>%
#   dplyr::group_by(Stage) %>%
#   dplyr::mutate(
#     prop = pStrata/sum(pStrata)
#   )
# 
# pStrata_data2 %>% 
#   ggplot2::ggplot(aes(x = Stage_number, y = prop, fill = PS)) +
#   ggplot2::geom_area(alpha=0.6 , size=1, colour="black")
```

```{r}
# # Works but uninteresting.
# pTAI_data <-
#   tibble::as_tibble(myTAI::pTAI(PhyloExpressionSetExample), rownames = "PS") %>%
#   tidyr::pivot_longer(!PS, names_to = "Stage", values_to = "pTAI_accumulated")
# 
# pTAI_data$Stage <-
#   pTAI_data$Stage <- base::factor(pTAI_data$Stage, unique(pTAI_data$Stage))
# 
# pTAI_data %>%
#   ggplot2::ggplot(aes(x = Stage, y = pTAI_accumulated, group = PS, colour = PS)) +
#   ggplot2::geom_line()
```

```{r}
PlotCategoryExpr(ExpressionSet = myTAI::tf(PhyloExpressionSetExample, FUN = log2, pseudocount = 1),
                     legendName    = "PS",
                     test.stat     = TRUE,
                     type          = "category-centered",
                     distr.type    = "boxplot",
                     log.expr      = FALSE)

PlotCategoryExpr(ExpressionSet = myTAI::tf(PhyloExpressionSetExample, FUN = log2, pseudocount = 1),
                     legendName    = "PS",
                     test.stat     = TRUE,
                     type          = "stage-centered",
                     distr.type    = "boxplot",
                     log.expr      = FALSE)
```

# Fucus analysis

Here, I focus on the stages that generates the _Fucus_ embryo, i.e. stages *not* `gamates` nor `matSP`.

```{r}
Fd_PES <-
  readr::read_csv(file = "data/Fd_PES.csv") %>%
  dplyr::select(!c("gamete","matSP"))
Fs_PES <-
  readr::read_csv(file = "data/Fs_PES.csv") %>%
  dplyr::select(!c("gamete","matSP"))
```

sqrt-tranformed

```{r}
Fd_PES.sqrt <-
  readr::read_csv(file = "data/Fd_PES.sqrt.csv") %>%
  dplyr::select(!c("gamete","matSP"))
Fs_PES.sqrt <-
  readr::read_csv(file = "data/Fs_PES.sqrt.csv") %>%
  dplyr::select(!c("gamete","matSP"))
```

log2-tranformed

```{r}
Fd_PES.log2 <-
  readr::read_csv(file = "data/Fd_PES.log2.csv") %>%
  dplyr::select(!c("gamete","matSP"))
Fs_PES.log2 <-
  readr::read_csv(file = "data/Fs_PES.log2.csv") %>%
  dplyr::select(!c("gamete","matSP"))

# different sexes
Fd_PES_sex.log2 <-
  readr::read_csv(file = "data/Fd_PES_M.log2.csv") %>%
  dplyr::rename(gamete_M = V1)
Fd_PES_sex.log2 <-
  readr::read_csv(file = "data/Fd_PES_F.log2.csv") %>%
  dplyr::rename(gamete_F = V1) %>%
  left_join(Fd_PES_sex.log2)
Fs_PES_sex.log2 <-
  readr::read_csv(file = "data/Fs_PES_M.log2.csv") %>% 
  dplyr::rename_with(
    ~ paste0(.x, "_M"),
    dplyr::starts_with(c("gamete", "mat")) &! dplyr::ends_with("_F"))
Fs_PES_sex.log2 <-
  readr::read_csv(file = "data/Fs_PES_F.log2.csv") %>% 
  dplyr::rename_with(
    ~ paste0(.x, "_F"),
    dplyr::starts_with(c("gamete", "mat")) &! dplyr::ends_with("_M")) %>%
  dplyr::left_join(Fs_PES_sex.log2)

Fs_PES_sex_mat.log2 <- 
  Fs_PES_sex.log2 %>%
  dplyr::select(!dplyr::starts_with("gamete"))
Fs_PES_sex_gam.log2 <- 
  Fs_PES_sex.log2 %>%
  dplyr::select(!dplyr::starts_with("mat"))

# different tissues
Fd_PES_matSP.log2 <-
  readr::read_csv(file = "data/Fd_PES_matSP.log2.csv")
Fs_PES_M_matSP.log2 <-
  readr::read_csv(file = "data/Fs_PES_M_matSP.log2.csv")
Fs_PES_F_matSP.log2 <-
  readr::read_csv(file = "data/Fs_PES_F_matSP.log2.csv")


Fs_PES_matSP_holdfast.log2 <-
  Fs_PES_M_matSP.log2 %>% 
  dplyr::select(1:2 | dplyr::starts_with("holdfast")) %>%
  dplyr::rename_with(
    ~ paste0(.x, "_M"),
    dplyr::starts_with(c("holdfast")) &! dplyr::ends_with("_F"))
Fs_PES_matSP_holdfast.log2 <-
  Fs_PES_F_matSP.log2 %>% 
  dplyr::select(1:2 | dplyr::starts_with("holdfast")) %>%
  dplyr::rename_with(
    ~ paste0(.x, "_F"),
    dplyr::starts_with(c("holdfast")) &! dplyr::ends_with("_M")) %>%
  dplyr::left_join(Fs_PES_matSP_holdfast.log2)

Fs_PES_matSP_reptip.log2 <-
  Fs_PES_M_matSP.log2 %>% 
  dplyr::select(1:2 | dplyr::starts_with("reptip")) %>%
  dplyr::rename_with(
    ~ paste0(.x, "_M"),
    dplyr::starts_with(c("reptip")) &! dplyr::ends_with("_F"))
Fs_PES_matSP_reptip.log2 <-
  Fs_PES_F_matSP.log2 %>% 
  dplyr::select(1:2 | dplyr::starts_with("reptip")) %>%
  dplyr::rename_with(
    ~ paste0(.x, "_F"),
    dplyr::starts_with(c("reptip")) &! dplyr::ends_with("_M")) %>%
  dplyr::left_join(Fs_PES_matSP_reptip.log2)

Fs_PES_matSP_stipe.log2 <-
  Fs_PES_M_matSP.log2 %>% 
  dplyr::select(1:2 | dplyr::starts_with("stipe")) %>%
  dplyr::rename_with(
    ~ paste0(.x, "_M"),
    dplyr::starts_with(c("stipe")) &! dplyr::ends_with("_F"))
Fs_PES_matSP_stipe.log2 <-
  Fs_PES_F_matSP.log2 %>% 
  dplyr::select(1:2 | dplyr::starts_with("stipe")) %>%
  dplyr::rename_with(
    ~ paste0(.x, "_F"),
    dplyr::starts_with(c("stipe")) &! dplyr::ends_with("_M")) %>%
  dplyr::left_join(Fs_PES_matSP_stipe.log2)

Fs_PES_matSP_vegtip.log2 <-
  Fs_PES_M_matSP.log2 %>% 
  dplyr::select(1:2 | dplyr::starts_with("vegtip")) %>%
  dplyr::rename_with(
    ~ paste0(.x, "_M"),
    dplyr::starts_with(c("vegtip")) &! dplyr::ends_with("_F"))
Fs_PES_matSP_vegtip.log2 <-
  Fs_PES_F_matSP.log2 %>% 
  dplyr::select(1:2 | dplyr::starts_with("vegtip")) %>%
  dplyr::rename_with(
    ~ paste0(.x, "_F"),
    dplyr::starts_with(c("vegtip")) &! dplyr::ends_with("_M")) %>%
  dplyr::left_join(Fs_PES_matSP_vegtip.log2)
```

rank-tranformed

```{r}
Fd_PES.rank <-
  readr::read_csv(file = "data/Fd_PES.rank.csv") %>%
  dplyr::select(!c("gamete","matSP"))
Fs_PES.rank <-
  readr::read_csv(file = "data/Fs_PES.rank.csv") %>%
  dplyr::select(!c("gamete","matSP"))
```

rlog-tranformed

```{r}
Fd_PES.rlog <-
  readr::read_csv(file = "data/Fd_PES.rlog.csv") %>%
  dplyr::select(!c("gamete","matSP"))
Fs_PES.rlog <-
  readr::read_csv(file = "data/Fs_PES.rlog.csv") %>%
  dplyr::select(!c("gamete","matSP"))
```

## PlotCategoryExpr()

### Fucus distichus

```{r}
# PlotCategoryExpr(
#   ExpressionSet = Fd_PES.log2,
#   legendName    = "PS",
#   test.stat     = TRUE,
#   type          = "category-centered",
#   distr.type    = "boxplot",
#   log.expr      = FALSE) +
#   ggplot2::labs(
#     title = "Fucus distichus",
#     subtitle = "log2(TPM+1)"
#   )

PlotCategoryExpr(
  ExpressionSet = Fd_PES.log2,
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Fucus distichus development",
    subtitle = "log2(TPM+1)"
  )

PlotCategoryExpr(
  ExpressionSet = Fd_PES_sex.log2,
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Fucus distichus gamete sexes",
    subtitle = "log2(TPM+1)"
  )
```

```{r}
PlotCategoryExpr(
  ExpressionSet = Fd_PES_matSP.log2,
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Fucus distichus matSP tissues",
    subtitle = "log2(TPM+1)"
  )
```

### Fucus serratus

```{r}
# PlotCategoryExpr(
#   ExpressionSet = Fs_PES.log2,
#   legendName    = "PS",
#   test.stat     = TRUE,
#   type          = "category-centered",
#   distr.type    = "boxplot",
#   log.expr      = FALSE) +
#   ggplot2::labs(
#     title = "Fucus serratus",
#     subtitle = "log2(TPM+1)"
#   )

PlotCategoryExpr(
  ExpressionSet = Fs_PES.log2,
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Fucus serratus development",
    subtitle = "log2(TPM+1)"
  )

PlotCategoryExpr(
  ExpressionSet = Fs_PES_sex_gam.log2,
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Fucus serratus gamete sexes",
    subtitle = "log2(TPM+1)"
  )

PlotCategoryExpr(
  ExpressionSet = Fs_PES_sex_mat.log2,
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Fucus serratus matSP sexes",
    subtitle = "log2(TPM+1)"
  )
```

```{r}
PlotCategoryExpr(
  ExpressionSet = Fs_PES_M_matSP.log2,
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Fucus serratus matSP (male) tissues",
    subtitle = "log2(TPM+1)"
  )

PlotCategoryExpr(
  ExpressionSet = Fs_PES_F_matSP.log2,
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Fucus serratus matSP (female) tissues",
    subtitle = "log2(TPM+1)"
  )
```

```{r}
PlotCategoryExpr(
  ExpressionSet = Fs_PES_matSP_vegtip.log2,
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Fucus serratus matSP vegtip",
    subtitle = "log2(TPM+1)"
  )

PlotCategoryExpr(
  ExpressionSet = Fs_PES_matSP_reptip.log2,
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Fucus serratus matSP reptip",
    subtitle = "log2(TPM+1)"
  )

PlotCategoryExpr(
  ExpressionSet = Fs_PES_matSP_stipe.log2,
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Fucus serratus matSP stipe",
    subtitle = "log2(TPM+1)"
  )

PlotCategoryExpr(
  ExpressionSet = Fs_PES_matSP_holdfast.log2,
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Fucus serratus matSP holdfast",
    subtitle = "log2(TPM+1)"
  )
```

The results are a lot clearer with our RNA-seq data compared to the microarray data of PhyloExpressionSetExample.

# Ectocarpus analysis

Here, I focus on the stages that generates the _Fucus_ embryo, i.e. stages *not* `gamates` nor `matSP`.

```{r}
Ec_PES_32m <-
  readr::read_csv(file = "data/Ec_PES_32m.csv")
Ec_PES_25f <-
  readr::read_csv(file = "data/Ec_PES_25f.csv")
```

sqrt-tranformed

```{r}
Ec_PES_32m.sqrt <-
  readr::read_csv(file = "data/Ec_PES_32m.sqrt.csv")
Ec_PES_25f.sqrt <-
  readr::read_csv(file = "data/Ec_PES_25f.sqrt.csv")
```

log2-tranformed

```{r}
Ec_PES_32m.log2 <-
  readr::read_csv(file = "data/Ec_PES_32m.log2.csv")
Ec_PES_25f.log2 <-
  readr::read_csv(file = "data/Ec_PES_25f.log2.csv")
```

rank-tranformed

```{r}
Ec_PES_32m.rank <-
  readr::read_csv(file = "data/Ec_PES_32m.rank.csv")
Ec_PES_25f.rank <-
  readr::read_csv(file = "data/Ec_PES_25f.rank.csv")
```

rlog-tranformed

```{r}
Ec_PES_32m.rlog <-
  readr::read_csv(file = "data/Ec_PES_32m.rlog.csv")
Ec_PES_25f.rlog <-
  readr::read_csv(file = "data/Ec_PES_25f.rlog.csv")
```

## PlotCategoryExpr()

### Ectocaprus Male

```{r, fig.height = 4, fig.width = 15}
PlotCategoryExpr(
  ExpressionSet = Ec_PES_32m.log2,
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Ectocarpus (male)",
    subtitle = "log2(TPM+1)"
  )


PlotCategoryExpr(
  ExpressionSet = Ec_PES_25f.log2,
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Ectocarpus (female)",
    subtitle = "log2(TPM+1)"
  )

PlotCategoryExpr(
  ExpressionSet = Ec_PES_32m.rlog,
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Ectocarpus (male)",
    subtitle = "rlog(TPM+1)"
  )


PlotCategoryExpr(
  ExpressionSet = Ec_PES_25f.rlog,
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Ectocarpus (female)",
    subtitle = "rlog(TPM+1)"
  )
```

```{r, fig.height = 4, fig.width = 8}
PlotCategoryExpr(
  ExpressionSet = dplyr::select(Ec_PES_32m.log2, c(1,2) | dplyr::ends_with("PSP")),
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Ectocarpus (male) multicellular PSP",
    subtitle = "log2(TPM+1)"
  )


PlotCategoryExpr(
  ExpressionSet = dplyr::select(Ec_PES_25f.log2, c(1,2) | dplyr::ends_with("PSP")),
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Ectocarpus (female) multicellular PSP",
    subtitle = "log2(TPM+1)"
  )
```

```{r, fig.height = 4, fig.width = 8}
PlotCategoryExpr(
  ExpressionSet = dplyr::select(Ec_PES_32m.log2, c(1,2) | dplyr::ends_with("GA")),
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Ectocarpus (male) multicellular GA",
    subtitle = "log2(TPM+1)"
  )


PlotCategoryExpr(
  ExpressionSet = dplyr::select(Ec_PES_25f.log2, c(1,2) | dplyr::ends_with("GA")),
  legendName    = "PS",
  test.stat     = TRUE,
  type          = "stage-centered",
  distr.type    = "boxplot",
  log.expr      = FALSE) +
  ggplot2::labs(
    title = "Ectocarpus (female) multicellular GA",
    subtitle = "log2(TPM+1)"
  )
```

# Summary

In _Fucus_, PS1 and PS8 are clearly differentially expressed on average between developmental stages (as defined in myTAI's `PlotCategoryExpr()`). PS2 and PS7 also give some traces of differential expression. In _F. distichus_ gametes, males have a lower expression level across most PS's (significant in PS1, PS2, PS6, PS7 and PS8). In _F. serratus_ gametes, this is also the case but the differences are more pronounced (significant across all PS's). _F. serratus_ is dioecious and in the matSP, we find that only PS8 is significantly differentially regulated between sexes when averaging all constituent tissues together.

Between tissue types of matSP in both _Fucus_ species, essentially all gene age categories are differentially expressed. Focusing on the differences between sexes of matSP tissue types in the dioecious _F. serratus_, we see no significant differences between sexes in the vegetative tip (vegtip). PS8 is differentially expressed in the holdfast. PS1, PS2, PS7 and PS8 are differentially expressed in the stipe. As we would suspect, the sex-differences are greatest in the reproductive tip (reptip).

Together, this indicates that PS1 and PS8 genes are recruited to play a large role in the development of _Fucus_. PS8 plays a role in adult sex differentiation in dioecious _F. serratus_. In both _F. serratus_ and _F. distichus_, the gametes differentially expresses many PS's. Furthermore, it is surprising that most PS's are differentially expressed between tissues, while the number of differentially expressed PS's is more restricted during development. It is also surprising that many PS are differentially expressed between sexes in the stipe in the dioecious _F. serratus_, compared to other non-reproductive tissues such as the holdfast and vegtip. As we expect, differential PS expression is greatest between sexes in the reptip, whose cell-type composition foretells the differences in the gamete.

Compared to the age profile in _Fucus_, in _Ectocarpus_ all gene ages are differentially expressed between developmental stages, even when considering just the multicellular PSP or GA. This points to a more dynamic transcriptional landscape through development, which involves genes from all levels of sequence-space conservation. When including all stages together, we see that the unicellar stages have less expression across all PS, though this difference could be due to the lower RNA quality in meiospores and mitospores. Lastly, unlike _Ectocarpus_, in _Fucus_, we lack some PS's and most genes belong to PS1 and PS8. Thus, the resolution for this analysis is less fine in _Fucus_.

Get session info.

```{r}
devtools::session_info()
```