---
title: "4.0 OG visualisation"
output:
  html_document:
    fig_width: 6
    fig_height: 4
    df_print: paged
editor_options: 
  chunk_output_type: inline
---

# Orthogroup expression analysis

Instead of relying on one-to-one orthologues, we incorporate in-paralogues into _comparative transcriptomics_ analysis. As I have discussed in the article vignette of `myTAI` (see [here](https://drostlab.github.io/myTAI/articles/Phylostratigraphy.html)), this approach has some limitations, in that while not explicitely transferring (potential) issues from the `orthologue conjecture`, some elements of this assumption is incoporated. Regardless, I think it still captures more information than using one-to-one orthologues.

This is different to the original filter because with the original filter, we wanted to retain as much genes as possible for TAI analysis. Here, we want to minimise species difference.

```{r}
library(tidyverse)
library(ggfortify)
```

```{r}
sample_info_fd <-
  readr::read_csv("data/sample_info_fd.csv")
sample_info_fs <-
  readr::read_csv("data/sample_info_fs.csv")
sample_info_ec <-
  readr::read_csv("data/sample_info_ec.csv")

sample_info_fd_embryo <-
  sample_info_fd %>%
  dplyr::filter(Stage %in% c("E1", "E2", "E3", "E4", "E5", "E6")) %>%
  dplyr::mutate(LibLab = str_c(Species, "_", Stage, "_", Replicate))
sample_info_fs_embryo <-
  sample_info_fs %>%
  dplyr::filter(Stage %in% c("24H", "48H", "1w", "3w", "4w")) %>%
  dplyr::mutate(Replicate = case_when(
    LibName == "Fs_X_zygotes_48h_1" ~ 4,
    LibName == "Fs_X_zygotes_48h_2" ~ 5,
    LibName == "Fs_X_zygotes_48h_3" ~ 6,
    TRUE ~ Replicate
  )) %>%
  mutate(LibLab = str_c(Species, "_", Stage, "_", Replicate))
```

```{r}
Fs_abundance <-
  readr::read_csv(file = "data/Fs_OG_abundance.csv")
Fs_counts <-
  readr::read_csv(file = "data/Fs_OG_counts.csv")
Fd_abundance <-
  readr::read_csv(file = "data/Fd_OG_abundance.csv")
Fd_counts <-
  readr::read_csv(file = "data/Fd_OG_counts.csv")
Ec_abundance <-
  readr::read_csv(file = "data/Ec_OG_abundance.csv")
Ec_counts <-
  readr::read_csv(file = "data/Ec_OG_counts.csv")
```


# PCA using OGs

```{r}
selectGenes <- function(counts, min.count=2){
  keep <-
    counts[rowMeans(counts)>min.count, ]
  return(keep)
}
```

To minimise differences between species, I remove OGs with low counts.

```{r}
combined_metatable <- 
  rbind(sample_info_fs_embryo, sample_info_fd_embryo) %>%
  relocate(LibName)
```

```{r}
merged_TPM_pre <- 
  merge(Fs_abundance, Fd_abundance, by = "OG") %>%
  dplyr::select(1, combined_metatable$LibName)
merged_TPM <- 
  data.frame(row.names = merged_TPM_pre[,1], merged_TPM_pre[,-1], check.names = FALSE)
merged_TPM.filt <-
  as.matrix(merged_TPM) %>%
  selectGenes(min.count=10)
```

Through the filtering function, I retain 2889 / 3724 OGs

```{r}
rlog_TPM <- DESeq2::rlog(merged_TPM.filt %>% round(0))
log2_TPM <- log2(merged_TPM.filt+1)
sqrt_TPM <- sqrt(merged_TPM.filt)
tpm10_TPM <- (merged_TPM.filt > 10) * 1
```

```{r fig.height = 8, fig.width = 5.5}
combined_metatable$Stage <- factor(combined_metatable$Stage, levels = unique(combined_metatable$Stage))

myPCAfunction <- function(pcDat, data, x, y){
  OUT <- 
    ggplot2::autoplot(
      pcDat,
      data = data,
      frame.colour = "Stage",
      colour = "Stage",
      shape = "Species",
      x = x,
      y = y,
      size = 3,
      frame = TRUE,
      alpha = 0.5)
  return(OUT)
}

pcDat <- prcomp(t(rlog_TPM), scale = TRUE)
# pcDat <- prcomp(t(rlog_TPM))

PC1_2 <- 
  myPCAfunction(pcDat, data = combined_metatable, x = 1, y = 2) +
  labs(title = "PC1 and PC2") +
  theme_grey()

PC1_3 <- 
  myPCAfunction(pcDat, data = combined_metatable, x = 1, y = 3) +
  labs(title = "PC1 and PC3") +
  theme_grey()

rlog_plot <- ggpubr::ggarrange(PC1_2, PC1_3, common.legend = T, ncol = 1, nrow = 2)
ggpubr::annotate_figure(rlog_plot, top = ggpubr::text_grob("rlog", 
                                                   face = "bold", size = 14))
```

```{r fig.height = 3, fig.width = 5.5}

pcDat <- prcomp(t(log2_TPM), scale = TRUE)

PC1_2 <- myPCAfunction(pcDat, data = combined_metatable, x = 1, y = 2) +
  labs(title = "PC1 and PC2") +
  theme_grey()

PC1_3 <- myPCAfunction(pcDat, data = combined_metatable, x = 1, y = 3) +
  labs(title = "PC1 and PC3") +
  theme_grey()

log2_plot <- ggpubr::ggarrange(PC1_2, PC1_3, common.legend = T)
ggpubr::annotate_figure(log2_plot, top = ggpubr::text_grob("log2", 
                                                   face = "bold", size = 14))

pcDat <- prcomp(t(sqrt_TPM), scale = TRUE)
# pcDat <- prcomp(t(rlog_TPM))

PC1_2 <- myPCAfunction(pcDat, data = combined_metatable, x = 1, y = 2) +
  labs(title = "PC1 and PC2") +
  theme_grey()

PC1_3 <- myPCAfunction(pcDat, data = combined_metatable, x = 1, y = 3) +
  labs(title = "PC1 and PC3") +
  theme_grey()

sqrt_plot <- ggpubr::ggarrange(PC1_2, PC1_3, common.legend = T)
ggpubr::annotate_figure(sqrt_plot, top = ggpubr::text_grob("sqrt", 
                                                   face = "bold", size = 14))

pcDat <- prcomp(t(tpm10_TPM)) # too many zeros to do a scale = TRUE

PC1_2 <- myPCAfunction(pcDat, data = combined_metatable, x = 1, y = 2) +
  labs(title = "PC1 and PC2") +
  theme_grey()

PC1_3 <- myPCAfunction(pcDat, data = combined_metatable, x = 1, y = 3) +
  labs(title = "PC1 and PC3") +
  theme_grey()

rlog_plot <- ggpubr::ggarrange(PC1_2, PC1_3, common.legend = T)
ggpubr::annotate_figure(rlog_plot, top = ggpubr::text_grob("tpm10 threshold", 
                                                   face = "bold", size = 14))
```

# Distance metrics

Here, we use distance metrics to quantify how distant the stages are from each other across _Fucus_ species and maybe even in _Ectocarpus_.

```{r}
library(philentropy)
```

```{r}
ncol_Fs <- grep( "Fs" , colnames( log2_TPM ) )
ncol_Fd <- grep( "Fd" , colnames( log2_TPM ) )
```

## Using log2(TPM+1)

```{r}
log2_TPM_renamed <- log2_TPM
base::colnames(log2_TPM_renamed) <- combined_metatable$LibLab
```

### Pearson correlation

Linear relationship

```{r fig.height = 4, fig.width = 5}
M = cor(x = log2_TPM_renamed[,ncol_Fs], 
        y = log2_TPM_renamed[,ncol_Fd], 
        method = "pearson")

M_melt <- reshape2::melt(M)
ggplot2::ggplot(M_melt, aes(x = Var2, y = Var1, fill = value)) +
  ggplot2::geom_tile() +
  ggplot2::theme_minimal() +
  ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ggplot2::scale_fill_gradient(low = "white", high = "#DC0000FF") +
  ggplot2::labs(
    x = "Fucus distichus samples",
    y = "Fucus serratus samples",
    title = "Pearson correlation (log2)",
    fill = "Correlation") +
  ggplot2::coord_flip()
```

```{r fig.height = 3, fig.width = 4.5}
tmp1 <- sample_info_fd_embryo %>% 
  select(Var2 = LibLab, Stage) %>% 
  full_join(M_melt, multiple = "all") %>%
  dplyr::rename(Fd = Stage)

tmp2 <- sample_info_fs_embryo %>% 
  select(Var1 = LibLab, Stage) %>% 
  full_join(tmp1, multiple = "all") %>%
  dplyr::rename(Fs = Stage)

tmp2$Fs <- factor(tmp2$Fs, levels = unique(tmp2$Fs))
tmp2$Fd <- factor(tmp2$Fd, levels = unique(tmp2$Fd))
tmp2 %>% group_by(Fs, Fd) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fs, y = Fd, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,3))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#DC0000FF") +
  labs(x = "Fucus serratus samples",
       y = "Fucus distichus samples",
       title = "Pearson correlation (log2)",
       fill = "median(correlation)")
```

### Spearman correlation

Monotonic relationship

```{r fig.height = 4, fig.width = 5}
log2_TPM_renamed <- log2_TPM
base::colnames(log2_TPM_renamed) <- combined_metatable$LibLab

M = cor(x = log2_TPM_renamed[,ncol_Fs], 
        y = log2_TPM_renamed[,ncol_Fd], 
        method = "spearman")

M_melt <- reshape2::melt(M)
ggplot2::ggplot(M_melt, aes(x = Var2, y = Var1, fill = value)) +
  ggplot2::geom_tile() +
  ggplot2::theme_minimal() +
  ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ggplot2::scale_fill_gradient(low = "white", high = "#DC0000FF") +
  ggplot2::labs(
    x = "Fucus distichus samples",
    y = "Fucus serratus samples",
    title = "Spearman correlation (log2TPM)",
    fill = "Correlation") +
  ggplot2::coord_flip()
```

```{r fig.height = 3, fig.width = 4.5}
tmp1 <- sample_info_fd_embryo %>% 
  select(Var2 = LibLab, Stage) %>% 
  full_join(M_melt, multiple = "all") %>%
  dplyr::rename(Fd = Stage)

tmp2 <- sample_info_fs_embryo %>% 
  select(Var1 = LibLab, Stage) %>% 
  full_join(tmp1, multiple = "all") %>%
  dplyr::rename(Fs = Stage)

tmp2$Fs <- factor(tmp2$Fs, levels = unique(tmp2$Fs))
tmp2$Fd <- factor(tmp2$Fd, levels = unique(tmp2$Fd))
tmp2 %>% group_by(Fs, Fd) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fs, y = Fd, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,3))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#DC0000FF") +
  labs(x = "Fucus serratus samples",
       y = "Fucus distichus samples",
       title = "Spearman correlation (log2)",
       fill = "median(correlation)")
```

### Manhattan distance

Manhattan distance is a distance metric used to measure the distance between two points in a grid-like system, such as a city block or a chessboard. It is calculated as the sum of the absolute differences between the coordinates of the two points. The term “Manhattan” is used because the grid-like structure and the right-angled turns of the streets in Manhattan, New York, resemble a chessboard. The Manhattan distance is also known as the L1 norm or taxicab distance. It is commonly used in machine learning and data science to measure similarity between data points or to cluster data.

It is L_{1} norm unlike Euclidean L_{2} norm.

Because it uses absolute values instead of exponentiation and rooting, it is said to be *more robust to outliers* compared to the euclidean distance. Additionally, the modulus is much faster to compute than exponentiation. It is a special case of the Minkowski distance.

```{r fig.height = 4, fig.width = 5}
DM <- philentropy::distance(t(log2_TPM_renamed), use.row.names = TRUE, method = "manhattan")
DM2 <- DM[ncol_Fs, ncol_Fd]

M_melt <- reshape2::melt(DM2)
ggplot2::ggplot(M_melt, aes(x = Var2, y = Var1, fill = value)) +
  ggplot2::geom_tile() +
  ggplot2::theme_minimal() +
  ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ggplot2::scale_fill_gradient(low = "white", high = "#3C5488FF") +
  ggplot2::labs(
    x = "Fucus distichus samples",
    y = "Fucus serratus samples",
    title = "Manhattan distance (log2)",
    fill = "Distance") +
  ggplot2::coord_flip()
```

```{r fig.height = 3, fig.width = 4.5}
tmp1 <- sample_info_fd_embryo %>% 
  select(Var2 = LibLab, Stage) %>% 
  full_join(M_melt, multiple = "all") %>%
  dplyr::rename(Fd = Stage)

tmp2 <- sample_info_fs_embryo %>% 
  select(Var1 = LibLab, Stage) %>% 
  full_join(tmp1, multiple = "all") %>%
  dplyr::rename(Fs = Stage)

tmp2$Fs <- factor(tmp2$Fs, levels = unique(tmp2$Fs))
tmp2$Fd <- factor(tmp2$Fd, levels = unique(tmp2$Fd))
tmp2 %>% group_by(Fs, Fd) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fs, y = Fd, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,0))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#3C5488FF") +
  labs(x = "Fucus serratus samples",
       y = "Fucus distichus samples",
       title = "Manhattan distance (log2)",
       fill = "median(distance)")
```

### Euclidean distance

Euclidean distance is a measure of the straight-line distance between two points in a multi-dimensional space. It is calculated as the square root of the sum of the squared differences between the corresponding coordinates of the two points. The Euclidean distance is commonly used in clustering and classification algorithms, as well as in distance-based data analysis and visualization techniques.

Euclidean is [not ideal for high-dimension data](https://stats.stackexchange.com/questions/99171/why-is-euclidean-distance-not-a-good-metric-in-high-dimensions/99191#99191). But it is commonly used.

```{r fig.height = 4, fig.width = 5}
DM <- philentropy::distance(t(log2_TPM_renamed), use.row.names = TRUE, method = "euclidean")
DM2 <- DM[ncol_Fs, ncol_Fd]

M_melt <- reshape2::melt(DM2)
ggplot2::ggplot(M_melt, aes(x = Var2, y = Var1, fill = value)) +
  ggplot2::geom_tile() +
  ggplot2::theme_minimal() +
  ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ggplot2::scale_fill_gradient(low = "white", high = "#3C5488FF") +
  ggplot2::labs(
    x = "Fucus distichus samples",
    y = "Fucus serratus samples",
    title = "Euclidean distance (log2)",
    fill = "Distance") +
  ggplot2::coord_flip()
```

```{r fig.height = 3, fig.width = 4.5}
tmp1 <- sample_info_fd_embryo %>% 
  select(Var2 = LibLab, Stage) %>% 
  full_join(M_melt, multiple = "all") %>%
  dplyr::rename(Fd = Stage)

tmp2 <- sample_info_fs_embryo %>% 
  select(Var1 = LibLab, Stage) %>% 
  full_join(tmp1, multiple = "all") %>%
  dplyr::rename(Fs = Stage)

tmp2$Fs <- factor(tmp2$Fs, levels = unique(tmp2$Fs))
tmp2$Fd <- factor(tmp2$Fd, levels = unique(tmp2$Fd))
tmp2 %>% group_by(Fs, Fd) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fs, y = Fd, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,1))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#3C5488FF") +
  labs(x = "Fucus serratus samples",
       y = "Fucus distichus samples",
       title = "Euclidean distance (log2)",
       fill = "median(distance)")
```

### JSD metric

Jensen-Shannon distance is a statistical distance metric that measures the similarity between two probability distributions. It is often used in data analysis and machine learning for clustering, classification, and anomaly detection tasks.

```{r fig.height = 4, fig.width = 5}
mat_normalized <- apply(log2_TPM_renamed, 2, function(x) x/sum(x)) %>% as.data.frame()
# colSums(mat_normalized)
DM <- philentropy::distance(t(mat_normalized), use.row.names = TRUE, method = "jensen-shannon")
DM2 <- DM[ncol_Fs, ncol_Fd]
DM3 <- sqrt(DM2)

M_melt <- reshape2::melt(DM3)
ggplot(M_melt, aes(x = Var2, y = Var1, fill = value)) +
  geom_tile() +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#3C5488FF") +
  labs(x = "Fucus distichus samples",
       y = "Fucus serratus samples",
       title = "JSD metric (norm. log2)",
       fill = "Distance") +
  coord_flip()
```


```{r fig.height = 3, fig.width = 4.5}
tmp1 <- sample_info_fd_embryo %>% 
  select(Var2 = LibLab, Stage) %>% 
  full_join(M_melt, multiple = "all") %>%
  dplyr::rename(Fd = Stage)

tmp2 <- sample_info_fs_embryo %>% 
  select(Var1 = LibLab, Stage) %>% 
  full_join(tmp1, multiple = "all") %>%
  dplyr::rename(Fs = Stage)

tmp2$Fs <- factor(tmp2$Fs, levels = unique(tmp2$Fs))
tmp2$Fd <- factor(tmp2$Fd, levels = unique(tmp2$Fd))
tmp2 %>% group_by(Fs, Fd) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fs, y = Fd, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,3))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#3C5488FF") +
  labs(x = "Fucus serratus samples",
       y = "Fucus distichus samples",
       title = "JSD metric (norm. log2)",
       fill = "median(distance)")
```

## Using rlog(TPM)

I am using rlog data here because it reduces the distance since there is higher correlation between the low-count genes. It should be noted though that the transformation here isn't monotonic.

```{r}
rlog_TPM_renamed <- rlog_TPM
rlog_TPM_renamed <- rlog_TPM
base::colnames(rlog_TPM_renamed) <- combined_metatable$LibLab
```

### Pearson correlation

```{r fig.height = 4, fig.width = 5}
base::colnames(rlog_TPM_renamed) <- combined_metatable$LibLab

M = cor(x = rlog_TPM_renamed[,ncol_Fs], 
        y = rlog_TPM_renamed[,ncol_Fd], 
        method = "pearson")

M_melt <- reshape2::melt(M)
ggplot2::ggplot(M_melt, aes(x = Var2, y = Var1, fill = value)) +
  ggplot2::geom_tile() +
  ggplot2::theme_minimal() +
  ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ggplot2::scale_fill_gradient(low = "white", high = "#DC0000FF") +
  ggplot2::labs(
    x = "Fucus distichus samples",
    y = "Fucus serratus samples",
    title = "Pearson correlation (rlog)",
    fill = "Correlation") +
  ggplot2::coord_flip()
```

```{r fig.height = 3, fig.width = 4.5}
tmp1 <- sample_info_fd_embryo %>% 
  select(Var2 = LibLab, Stage) %>% 
  full_join(M_melt, multiple = "all") %>%
  dplyr::rename(Fd = Stage)

tmp2 <- sample_info_fs_embryo %>% 
  select(Var1 = LibLab, Stage) %>% 
  full_join(tmp1, multiple = "all") %>%
  dplyr::rename(Fs = Stage)

tmp2$Fs <- factor(tmp2$Fs, levels = unique(tmp2$Fs))
tmp2$Fd <- factor(tmp2$Fd, levels = unique(tmp2$Fd))
tmp2 %>% group_by(Fs, Fd) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fs, y = Fd, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,3))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#DC0000FF") +
  labs(x = "Fucus serratus samples",
       y = "Fucus distichus samples",
       title = "Pearson correlation (rlog)",
       fill = "median(correlation)")
```

### Spearman correlation

```{r fig.height = 4, fig.width = 5}
rlog_TPM_renamed <- rlog_TPM
base::colnames(rlog_TPM_renamed) <- combined_metatable$LibLab

M = cor(x = rlog_TPM_renamed[,ncol_Fs], 
        y = rlog_TPM_renamed[,ncol_Fd], 
        method = "spearman")

M_melt <- reshape2::melt(M)
ggplot2::ggplot(M_melt, aes(x = Var2, y = Var1, fill = value)) +
  ggplot2::geom_tile() +
  ggplot2::theme_minimal() +
  ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ggplot2::scale_fill_gradient(low = "white", high = "#DC0000FF") +
  ggplot2::labs(
    x = "Fucus distichus samples",
    y = "Fucus serratus samples",
    title = "Spearman correlation (rlog)",
    fill = "Correlation") +
  ggplot2::coord_flip()
```

```{r fig.height = 3, fig.width = 4.5}
tmp1 <- sample_info_fd_embryo %>% 
  select(Var2 = LibLab, Stage) %>% 
  full_join(M_melt, multiple = "all") %>%
  dplyr::rename(Fd = Stage)

tmp2 <- sample_info_fs_embryo %>% 
  select(Var1 = LibLab, Stage) %>% 
  full_join(tmp1, multiple = "all") %>%
  dplyr::rename(Fs = Stage)

tmp2$Fs <- factor(tmp2$Fs, levels = unique(tmp2$Fs))
tmp2$Fd <- factor(tmp2$Fd, levels = unique(tmp2$Fd))
tmp2 %>% group_by(Fs, Fd) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fs, y = Fd, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,3))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#DC0000FF") +
  labs(x = "Fucus serratus samples",
       y = "Fucus distichus samples",
       title = "Spearman correlation (rlog)",
       fill = "median(correlation)")
```

### Manhattan distance

```{r fig.height = 4, fig.width = 5}
DM <- philentropy::distance(t(rlog_TPM_renamed), use.row.names = TRUE, method = "manhattan")
DM2 <- DM[ncol_Fs, ncol_Fd]

M_melt <- reshape2::melt(DM2)
ggplot2::ggplot(M_melt, aes(x = Var2, y = Var1, fill = value)) +
  ggplot2::geom_tile() +
  ggplot2::theme_minimal() +
  ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ggplot2::scale_fill_gradient(low = "white", high = "#3C5488FF") +
  ggplot2::labs(
    x = "Fucus distichus samples",
    y = "Fucus serratus samples",
    title = "Manhattan distance (rlog)",
    fill = "Distance") +
  ggplot2::coord_flip()
```

```{r fig.height = 3, fig.width = 4.5}
tmp1 <- sample_info_fd_embryo %>% 
  select(Var2 = LibLab, Stage) %>% 
  full_join(M_melt, multiple = "all") %>%
  dplyr::rename(Fd = Stage)

tmp2 <- sample_info_fs_embryo %>% 
  select(Var1 = LibLab, Stage) %>% 
  full_join(tmp1, multiple = "all") %>%
  dplyr::rename(Fs = Stage)

tmp2$Fs <- factor(tmp2$Fs, levels = unique(tmp2$Fs))
tmp2$Fd <- factor(tmp2$Fd, levels = unique(tmp2$Fd))
tmp2 %>% group_by(Fs, Fd) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fs, y = Fd, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,0))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#3C5488FF") +
  labs(x = "Fucus serratus samples",
       y = "Fucus distichus samples",
       title = "Manhattan distance (rlog)",
       fill = "median(distance)")
```

### Euclidean distance

```{r fig.height = 4, fig.width = 5}
DM <- philentropy::distance(t(rlog_TPM_renamed), use.row.names = TRUE, method = "euclidean")
DM2 <- DM[ncol_Fs, ncol_Fd]

M_melt <- reshape2::melt(DM2)
ggplot2::ggplot(M_melt, aes(x = Var2, y = Var1, fill = value)) +
  ggplot2::geom_tile() +
  ggplot2::theme_minimal() +
  ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ggplot2::scale_fill_gradient(low = "white", high = "#3C5488FF") +
  ggplot2::labs(
    x = "Fucus distichus samples",
    y = "Fucus serratus samples",
    title = "Euclidean distance (rlog)",
    fill = "Distance") +
  ggplot2::coord_flip()
```
```{r fig.height = 3, fig.width = 4.5}
tmp1 <- sample_info_fd_embryo %>% 
  select(Var2 = LibLab, Stage) %>% 
  full_join(M_melt, multiple = "all") %>%
  dplyr::rename(Fd = Stage)

tmp2 <- sample_info_fs_embryo %>% 
  select(Var1 = LibLab, Stage) %>% 
  full_join(tmp1, multiple = "all") %>%
  dplyr::rename(Fs = Stage)

tmp2$Fs <- factor(tmp2$Fs, levels = unique(tmp2$Fs))
tmp2$Fd <- factor(tmp2$Fd, levels = unique(tmp2$Fd))
tmp2 %>% group_by(Fs, Fd) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fs, y = Fd, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,1))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#3C5488FF") +
  labs(x = "Fucus serratus samples",
       y = "Fucus distichus samples",
       title = "Euclidean distance (rlog)",
       fill = "median(distance)")
```

### JSD metric

```{r fig.height = 4, fig.width = 5}
mat_normalized <- apply(rlog_TPM_renamed, 2, function(x) x/sum(x)) %>% as.data.frame()
# colSums(mat_normalized)
DM <- philentropy::distance(t(mat_normalized), use.row.names = TRUE, method = "jensen-shannon")
DM2 <- DM[ncol_Fs, ncol_Fd]
DM3 <- sqrt(DM2)

M_melt <- reshape2::melt(DM3)
ggplot(M_melt, aes(x = Var2, y = Var1, fill = value)) +
  geom_tile() +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#3C5488FF") +
  labs(x = "Fucus distichus samples",
       y = "Fucus serratus samples",
       title = "JSD metric (norm. rlog)",
       fill = "Distance") +
  coord_flip()
```


```{r fig.height = 3, fig.width = 4.5}
tmp1 <- sample_info_fd_embryo %>% 
  select(Var2 = LibLab, Stage) %>% 
  full_join(M_melt, multiple = "all") %>%
  dplyr::rename(Fd = Stage)

tmp2 <- sample_info_fs_embryo %>% 
  select(Var1 = LibLab, Stage) %>% 
  full_join(tmp1, multiple = "all") %>%
  dplyr::rename(Fs = Stage)

tmp2$Fs <- factor(tmp2$Fs, levels = unique(tmp2$Fs))
tmp2$Fd <- factor(tmp2$Fd, levels = unique(tmp2$Fd))
tmp2 %>% group_by(Fs, Fd) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fs, y = Fd, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,3))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#3C5488FF") +
  labs(x = "Fucus serratus samples",
       y = "Fucus distichus samples",
       title = "JSD metric (norm. rlog)",
       fill = "median(distance)")
```



## Compared to Ectocarpus?

Here we use JSD metric.

```{r}
sample_info_ec_proc <-
  sample_info_ec %>%
  dplyr::relocate(LibName) %>%
  dplyr::mutate(LibLab = str_c(Species, "_", Sex, "_", Stage, "_", Replicate))

duplicated(sample_info_ec_proc$LibLab)

combined_metatable_inclEc <- 
  rbind(combined_metatable, sample_info_ec_proc)
```

```{r}
merged_TPM_inclEc_pre <- 
  Ec_abundance %>%
  dplyr::select(1, sample_info_ec_proc$SampleName)

merged_TPM_inclEc_pre <-
  merged_TPM_pre %>%
  merge(merged_TPM_inclEc_pre)
merged_TPM_inclEc <- 
  data.frame(row.names = merged_TPM_inclEc_pre[,1], merged_TPM_inclEc_pre[,-1], check.names = FALSE)

merged_TPM.filt <-
  as.matrix(merged_TPM_inclEc) %>%
  selectGenes(min.count=10)
```

We keep 1620 OGs out of 1979 OGs.

```{r}
log2_TPM <- log2(merged_TPM.filt + 1)
```

```{r}
log2_TPM_renamed <- log2_TPM
base::colnames(log2_TPM_renamed) <- combined_metatable_inclEc$LibLab
```

```{r}
ncol_Fs <- grep( "Fs" , colnames( log2_TPM_renamed ) )
ncol_Fd <- grep( "Fd" , colnames( log2_TPM_renamed ) )
ncol_Ec <- grep( "Ec_" , colnames( log2_TPM_renamed ) )
```

### Pearson correlation

#### Fs & Ec

```{r fig.height = 10, fig.width = 5}
M = cor(x = log2_TPM_renamed[,ncol_Fs], 
        y = log2_TPM_renamed[,ncol_Ec], 
        method = "pearson")

M_melt <- reshape2::melt(M)
ggplot2::ggplot(M_melt, aes(x = Var2, y = Var1, fill = value)) +
  ggplot2::geom_tile() +
  ggplot2::theme_minimal() +
  ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ggplot2::scale_fill_gradient(low = "white", high = "#DC0000FF") +
  ggplot2::labs(
    x = "Ectocarpus samples",
    y = "Fucus serratus samples",
    title = "Pearson correlation (log2)",
    fill = "Correlation") +
  ggplot2::coord_flip()
```

```{r}
tmp1 <- sample_info_ec_proc %>% 
  dplyr::select(Var2 = LibLab, Stage, Sex) %>% 
  dplyr::mutate(StageSex = str_c(Stage, "_", Sex)) %>%
  dplyr::full_join(M_melt, multiple = "all") %>%
  dplyr::rename(Ec = StageSex) %>%
  dplyr::select(-Stage, -Sex)

tmp2 <- sample_info_fs_embryo %>% 
  select(Var1 = LibLab, Stage) %>% 
  full_join(tmp1, multiple = "all") %>%
  dplyr::rename(Fs = Stage)

tmp2$Fs <- factor(tmp2$Fs, levels = unique(tmp2$Fs))
tmp2$Ec <- factor(tmp2$Ec, levels = unique(tmp2$Ec))
tmp2 %>% group_by(Fs, Ec) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fs, y = Ec, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,3))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#DC0000FF") +
  labs(x = "Fucus serratus samples",
       y = "Ectocarpus samples",
       title = "Pearson correlation (log2)",
       fill = "median(correlation)")

tmp2 %>% dplyr::filter(!stringr::str_detect(Ec, 'spore|gamete')) %>% group_by(Fs, Ec) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fs, y = Ec, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,3))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#DC0000FF") +
  labs(x = "Fucus serratus samples",
       y = "Ectocarpus samples",
       title = "Pearson correlation (log2) [only multicellular]",
       fill = "median(correlation)")
```

#### Fd & Ec

```{r fig.height = 10, fig.width = 5}
M = cor(x = log2_TPM_renamed[,ncol_Fd], 
        y = log2_TPM_renamed[,ncol_Ec], 
        method = "pearson")

M_melt <- reshape2::melt(M)

ggplot2::ggplot(M_melt, aes(x = Var2, y = Var1, fill = value)) +
  ggplot2::geom_tile() +
  ggplot2::theme_minimal() +
  ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ggplot2::scale_fill_gradient(low = "white", high = "#DC0000FF") +
  ggplot2::labs(
    x = "Ectocarpus samples",
    y = "Fucus distichus samples",
    title = "Pearson correlation (log2)",
    fill = "Correlation") +
  ggplot2::coord_flip()
```

```{r}
tmp1 <- sample_info_ec_proc %>% 
  dplyr::select(Var2 = LibLab, Stage, Sex) %>% 
  dplyr::mutate(StageSex = str_c(Stage, "_", Sex)) %>%
  dplyr::full_join(M_melt, multiple = "all") %>%
  dplyr::rename(Ec = StageSex) %>%
  dplyr::select(-Stage, -Sex)

tmp2 <- sample_info_fd_embryo %>% 
  select(Var1 = LibLab, Stage) %>% 
  full_join(tmp1, multiple = "all") %>%
  dplyr::rename(Fd = Stage)

tmp2$Fd <- factor(tmp2$Fd, levels = unique(tmp2$Fd))
tmp2$Ec <- factor(tmp2$Ec, levels = unique(tmp2$Ec))
tmp2 %>% group_by(Fd, Ec) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fd, y = Ec, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,3))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#DC0000FF") +
  labs(x = "Fucus distichus samples",
       y = "Ectocarpus samples",
       title = "Pearson correlation (log2)",
       fill = "median(correlation)")

tmp2 %>% dplyr::filter(!stringr::str_detect(Ec, 'spore|gamete'))  %>% group_by(Fd, Ec) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fd, y = Ec, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,3))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#DC0000FF") +
  labs(x = "Fucus distichus samples",
       y = "Ectocarpus samples",
       title = "Pearson correlation (log2) [only multicellular]",
       fill = "median(correlation)")
```

### Spearman correlation

#### Fs & Ec

```{r}
M = cor(x = log2_TPM_renamed[,ncol_Fs], 
        y = log2_TPM_renamed[,ncol_Ec], 
        method = "spearman")

M_melt <- reshape2::melt(M)

tmp1 <- sample_info_ec_proc %>% 
  dplyr::select(Var2 = LibLab, Stage, Sex) %>% 
  dplyr::mutate(StageSex = str_c(Stage, "_", Sex)) %>%
  dplyr::full_join(M_melt, multiple = "all") %>%
  dplyr::rename(Ec = StageSex) %>%
  dplyr::select(-Stage, -Sex)

tmp2 <- sample_info_fs_embryo %>% 
  select(Var1 = LibLab, Stage) %>% 
  full_join(tmp1, multiple = "all") %>%
  dplyr::rename(Fs = Stage)

tmp2$Fs <- factor(tmp2$Fs, levels = unique(tmp2$Fs))
tmp2$Ec <- factor(tmp2$Ec, levels = unique(tmp2$Ec))
tmp2 %>% group_by(Fs, Ec) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fs, y = Ec, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,3))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#DC0000FF") +
  labs(x = "Fucus serratus samples",
       y = "Ectocarpus samples",
       title = "Spearman correlation (log2)",
       fill = "median(correlation)")

tmp2 %>% dplyr::filter(!stringr::str_detect(Ec, 'spore|gamete')) %>% group_by(Fs, Ec) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fs, y = Ec, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,3))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#DC0000FF") +
  labs(x = "Fucus serratus samples",
       y = "Ectocarpus samples",
       title = "Spearman correlation (log2) [only multicellular]",
       fill = "median(correlation)")
```

#### Fd & Ec

```{r}
M = cor(x = log2_TPM_renamed[,ncol_Fd], 
        y = log2_TPM_renamed[,ncol_Ec], 
        method = "spearman")

M_melt <- reshape2::melt(M)

tmp1 <- sample_info_ec_proc %>% 
  dplyr::select(Var2 = LibLab, Stage, Sex) %>% 
  dplyr::mutate(StageSex = str_c(Stage, "_", Sex)) %>%
  dplyr::full_join(M_melt, multiple = "all") %>%
  dplyr::rename(Ec = StageSex) %>%
  dplyr::select(-Stage, -Sex)

tmp2 <- sample_info_fd_embryo %>% 
  select(Var1 = LibLab, Stage) %>% 
  full_join(tmp1, multiple = "all") %>%
  dplyr::rename(Fd = Stage)

tmp2$Fd <- factor(tmp2$Fd, levels = unique(tmp2$Fd))
tmp2$Ec <- factor(tmp2$Ec, levels = unique(tmp2$Ec))
tmp2 %>% group_by(Fd, Ec) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fd, y = Ec, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,3))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#DC0000FF") +
  labs(x = "Fucus distichus samples",
       y = "Ectocarpus samples",
       title = "Spearman correlation (log2)",
       fill = "median(correlation)")

tmp2 %>% dplyr::filter(!stringr::str_detect(Ec, 'spore|gamete')) %>% group_by(Fd, Ec) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fd, y = Ec, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,3))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#DC0000FF") +
  labs(x = "Fucus distichus samples",
       y = "Ectocarpus samples",
       title = "Spearman correlation (log2) [only multicellular]",
       fill = "median(correlation)")
```

### JSD metric

#### Fs & Ec

```{r}
mat_normalized <- apply(log2_TPM_renamed, 2, function(x) x/sum(x)) %>% as.data.frame()
# colSums(mat_normalized)
DM <- philentropy::distance(t(mat_normalized), use.row.names = TRUE, method = "jensen-shannon")
DM2 <- DM[ncol_Fs, ncol_Ec]
DM3 <- sqrt(DM2)

M_melt <- reshape2::melt(DM3)

tmp1 <- sample_info_ec_proc %>% 
  dplyr::select(Var2 = LibLab, Stage, Sex) %>% 
  dplyr::mutate(StageSex = str_c(Stage, "_", Sex)) %>%
  dplyr::full_join(M_melt, multiple = "all") %>%
  dplyr::rename(Ec = StageSex) %>%
  dplyr::select(-Stage, -Sex)

tmp2 <- sample_info_fs_embryo %>% 
  select(Var1 = LibLab, Stage) %>% 
  full_join(tmp1, multiple = "all") %>%
  dplyr::rename(Fs = Stage)

tmp2$Fs <- factor(tmp2$Fs, levels = unique(tmp2$Fs))
tmp2$Ec <- factor(tmp2$Ec, levels = unique(tmp2$Ec))
tmp2 %>% group_by(Fs, Ec) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fs, y = Ec, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,3))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#3C5488FF") +
  labs(x = "Fucus serratus samples",
       y = "Ectocarpus samples",
       title = "JSD metric (norm. log2)",
       fill = "median(distance)")

tmp2 %>% dplyr::filter(!stringr::str_detect(Ec, 'spore|gamete')) %>% group_by(Fs, Ec) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fs, y = Ec, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,3))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#3C5488FF") +
  labs(x = "Fucus serratus samples",
       y = "Ectocarpus samples",
       title = "JSD metric (norm. log2) [only multicellular]",
       fill = "median(distance)")
```

#### Fd & Ec

```{r}
DM2 <- DM[ncol_Fd, ncol_Ec]
DM3 <- sqrt(DM2)

M_melt <- reshape2::melt(DM3)

tmp1 <- sample_info_ec_proc %>% 
  dplyr::select(Var2 = LibLab, Stage, Sex) %>% 
  dplyr::mutate(StageSex = str_c(Stage, "_", Sex)) %>%
  dplyr::full_join(M_melt, multiple = "all") %>%
  dplyr::rename(Ec = StageSex) %>%
  dplyr::select(-Stage, -Sex)

tmp2 <- sample_info_fd_embryo %>% 
  select(Var1 = LibLab, Stage) %>% 
  full_join(tmp1, multiple = "all") %>%
  dplyr::rename(Fd = Stage)

tmp2$Fd <- factor(tmp2$Fd, levels = unique(tmp2$Fd))
tmp2$Ec <- factor(tmp2$Ec, levels = unique(tmp2$Ec))
tmp2 %>% group_by(Fd, Ec) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fd, y = Ec, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,3))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#3C5488FF") +
  labs(x = "Fucus distichus samples",
       y = "Ectocarpus samples",
       title = "JSD metric (norm. log2)",
       fill = "median(distance)")

tmp2 %>% dplyr::filter(!stringr::str_detect(Ec, 'spore|gamete')) %>% group_by(Fd, Ec) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fd, y = Ec, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,3))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#3C5488FF") +
  labs(x = "Fucus distichus samples",
       y = "Ectocarpus samples",
       title = "JSD metric (norm. log2) [only multicellular]",
       fill = "median(distance)")
```

### Manhattan distance

#### Fs & Ec

```{r}
DM <- philentropy::distance(t(log2_TPM_renamed), use.row.names = TRUE, method = "manhattan")
DM2 <- DM[ncol_Fs, ncol_Ec]

M_melt <- reshape2::melt(DM2)

tmp1 <- sample_info_ec_proc %>% 
  dplyr::select(Var2 = LibLab, Stage, Sex) %>% 
  dplyr::mutate(StageSex = str_c(Stage, "_", Sex)) %>%
  dplyr::full_join(M_melt, multiple = "all") %>%
  dplyr::rename(Ec = StageSex) %>%
  dplyr::select(-Stage, -Sex)

tmp2 <- sample_info_fs_embryo %>% 
  select(Var1 = LibLab, Stage) %>% 
  full_join(tmp1, multiple = "all") %>%
  dplyr::rename(Fs = Stage)

tmp2$Fs <- factor(tmp2$Fs, levels = unique(tmp2$Fs))
tmp2$Ec <- factor(tmp2$Ec, levels = unique(tmp2$Ec))
tmp2 %>% group_by(Fs, Ec) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fs, y = Ec, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,0))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#3C5488FF") +
  labs(x = "Fucus serratus samples",
       y = "Ectocarpus samples",
       title = "Manhattan distance (log2)",
       fill = "median(distance)")

tmp2 %>% dplyr::filter(!stringr::str_detect(Ec, 'spore|gamete')) %>% group_by(Fs, Ec) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fs, y = Ec, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,0))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#3C5488FF") +
  labs(x = "Fucus serratus samples",
       y = "Ectocarpus samples",
       title = "Manhattan distance (log2) [only multicellular]",
       fill = "median(distance)")
```

#### Fd & Ec

```{r}
DM2 <- DM[ncol_Fd, ncol_Ec]

M_melt <- reshape2::melt(DM2)

tmp1 <- sample_info_ec_proc %>% 
  dplyr::select(Var2 = LibLab, Stage, Sex) %>% 
  dplyr::mutate(StageSex = str_c(Stage, "_", Sex)) %>%
  dplyr::full_join(M_melt, multiple = "all") %>%
  dplyr::rename(Ec = StageSex) %>%
  dplyr::select(-Stage, -Sex)

tmp2 <- sample_info_fd_embryo %>% 
  select(Var1 = LibLab, Stage) %>% 
  full_join(tmp1, multiple = "all") %>%
  dplyr::rename(Fd = Stage)

tmp2$Fd <- factor(tmp2$Fd, levels = unique(tmp2$Fd))
tmp2$Ec <- factor(tmp2$Ec, levels = unique(tmp2$Ec))
tmp2 %>% group_by(Fd, Ec) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fd, y = Ec, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,0))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#3C5488FF") +
  labs(x = "Fucus distichus samples",
       y = "Ectocarpus samples",
       title = "Manhattan distance (log2)",
       fill = "median(distance)")

tmp2 %>% dplyr::filter(!stringr::str_detect(Ec, 'spore|gamete')) %>% group_by(Fd, Ec) %>% summarise(median_corr = median(value)) %>%
  ggplot(aes(x = Fd, y = Ec, fill = median_corr)) +
  geom_tile() +
  geom_text(aes(label = paste0(round(median_corr,0))), color = "black") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_gradient(low = "white", high = "#3C5488FF") +
  labs(x = "Fucus distichus samples",
       y = "Ectocarpus samples",
       title = "Manhattan distance (log2) [only multicellular]",
       fill = "median(distance)")
```

Essentially, the difference between _Fucus_ species and _Ectocarpus_ is minimised during the stage of the lowest TAI.

Wesentlich wird der Unterschied zwischen _Fucus_-Arten und _Ectocarpus_ während der Phasen niedriges TAIs verringert.

# Males and female _Ectocarpus_ compared

## rlog transform


```{r}
Ec_PES_32m.rlog <-
        readr::read_csv(file = "data/Ec_PES_32m.rlog.csv") %>%
        dplyr::select(-1)

# Rename the columns with the prefix "M_"
new_colnames <- c("GeneID", paste0("M_", colnames(Ec_PES_32m.rlog)[-1]))
colnames(Ec_PES_32m.rlog) <- new_colnames


Ec_PES_25f.rlog <-
        readr::read_csv(file = "data/Ec_PES_25f.rlog.csv") %>%
        dplyr::select(-1)

# Rename the columns with the prefix "F_"
new_colnames <- c("GeneID", paste0("F_", colnames(Ec_PES_25f.rlog)[-1]))
colnames(Ec_PES_25f.rlog) <- new_colnames

Ec.rlog <- inner_join(Ec_PES_32m.rlog, Ec_PES_25f.rlog) %>%
        tibble::column_to_rownames(var = "GeneID")

ncol_Female <- grep( "F_" , colnames( Ec.rlog ) )
ncol_Male <- grep( "M_" , colnames( Ec.rlog ) )

```

#### Pearson correlation

```{r fig.height = 1.5, fig.width = 5}
M = cor(x = Ec.rlog[,ncol_Female],
        y = Ec.rlog[,ncol_Male],
        method = "pearson")

M_melt <- reshape2::melt(M) %>%
        dplyr::group_by(Var1, Var2) %>% 
        dplyr::summarise(median_corr = median(value)) %>%
        dplyr::filter(stringr::str_remove(Var2, "M_") == stringr::str_remove(Var1, "F_"))

M_melt_processed <- M_melt %>% mutate(Stage = stringr::str_remove(Var2, "M_"))
M_melt_processed$Stage <- factor(M_melt_processed$Stage, levels = unique(M_melt_processed$Stage))

ggplot2::ggplot(M_melt_processed, aes(x = Stage, y = 1, fill = median_corr, label = round(median_corr,2))) +
        ggplot2::geom_tile() +
        ggplot2::geom_text() +
        ggplot2::theme_minimal() +
        ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
                       legend.position = "",
                       axis.text.y=element_blank()) +
        ggplot2::scale_fill_gradient(low = "white", high = "#DC0000FF") +
        ggplot2::labs(
                x = "Ectocarpus stages",
                y = "Pearson",
                fill = "Correlation")
```

#### Spearman correlation

```{r fig.height = 1.5, fig.width = 5}
M = cor(x = Ec.rlog[,ncol_Female],
        y = Ec.rlog[,ncol_Male],
        method = "spearman")

M_melt <- reshape2::melt(M) %>%
        dplyr::group_by(Var1, Var2) %>% 
        dplyr::summarise(median_corr = median(value)) %>%
        dplyr::filter(stringr::str_remove(Var2, "M_") == stringr::str_remove(Var1, "F_"))

M_melt_processed <- M_melt %>% mutate(Stage = stringr::str_remove(Var2, "M_"))
M_melt_processed$Stage <- factor(M_melt_processed$Stage, levels = unique(M_melt_processed$Stage))

ggplot2::ggplot(M_melt_processed, aes(x = Stage, y = 1, fill = median_corr, label = round(median_corr,2))) +
        ggplot2::geom_tile() +
        ggplot2::geom_text() +
        ggplot2::theme_minimal() +
        ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
                       legend.position = "",
                       axis.text.y=element_blank()) +
        ggplot2::scale_fill_gradient(low = "white", high = "#DC0000FF") +
        ggplot2::labs(
                x = "Ectocarpus stages",
                y = "Spearman",
                fill = "Correlation")
```

#### JSD metric

Traditional Jensen-Shannon divergence (JSD) ranges between [0, 1]. Due to rlog transform creating less than 0 counts, I had to remove rows with at least one column with less than 0 counts since the normalisation step wouldn't work. This allows us to analyse 7159 genes instead of 11571 genes.

```{r}
library(matrixStats)

Ec.rlog_filt <- Ec.rlog[matrixStats::rowMins(as.matrix(Ec.rlog)) > 0, ]
mat_normalized <- apply(Ec.rlog_filt, 2, function(x) x/sum(x)) %>% as.data.frame()
DM <- philentropy::distance(t(mat_normalized), use.row.names = TRUE, method = "jensen-shannon")
DM2 <- sqrt(DM)
```

```{r fig.height = 1.5, fig.width = 5}
M_melt <- reshape2::melt(DM2) %>%
        dplyr::group_by(Var1, Var2) %>% 
        dplyr::summarise(median_corr = median(value)) %>%
        dplyr::filter(stringr::str_remove(Var2, "M_") == stringr::str_remove(Var1, "F_"))

M_melt_processed <- M_melt %>% mutate(Stage = stringr::str_remove(Var2, "M_"))
M_melt_processed$Stage <- factor(M_melt_processed$Stage, levels = unique(M_melt_processed$Stage))

ggplot2::ggplot(M_melt_processed, aes(x = Stage, y = 1, fill = median_corr, label = round(median_corr,3))) +
        ggplot2::geom_tile() +
        ggplot2::geom_text() +
        ggplot2::theme_minimal() +
        ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
                       legend.position = "",
                       axis.text.y=element_blank()) +
        ggplot2::scale_fill_gradient(low = "white", high = "#3C5488FF") +
        ggplot2::labs(
                x = "Ectocarpus stages",
                y = "JSD")
```

#### Manhattan

```{r}
DM <- philentropy::distance(t(Ec.rlog), use.row.names = TRUE, method = "manhattan")
```

```{r fig.height = 1.5, fig.width = 5}
M_melt <- reshape2::melt(DM) %>%
        dplyr::group_by(Var1, Var2) %>% 
        dplyr::summarise(median_corr = median(value)) %>%
        dplyr::filter(stringr::str_remove(Var2, "M_") == stringr::str_remove(Var1, "F_"))

M_melt_processed <- M_melt %>% mutate(Stage = stringr::str_remove(Var2, "M_"))
M_melt_processed$Stage <- factor(M_melt_processed$Stage, levels = unique(M_melt_processed$Stage))

ggplot2::ggplot(M_melt_processed, aes(x = Stage, y = 1, fill = median_corr, label = round(median_corr,0))) +
        ggplot2::geom_tile() +
        ggplot2::geom_text() +
        ggplot2::theme_minimal() +
        ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
                       legend.position = "",
                       axis.text.y=element_blank()) +
        ggplot2::scale_fill_gradient(low = "white", high = "#3C5488FF") +
        ggplot2::labs(
                x = "Ectocarpus stages",
                y = "Manhattan")
```

## log2(x+1) transform


```{r}
sample_info_ec_proc <-
        sample_info_ec %>%
        dplyr::relocate(LibName) %>% dplyr::filter(!Sex == "H") %>%
        dplyr::mutate(LibLab = str_c(Sex, "_", Stage))

Ec_abundance <-
        readr::read_csv(file = "data/Ec_abundance.csv") %>%
        dplyr::select(1, sample_info_ec_proc$SampleName) %>%
        tibble::column_to_rownames(var = "GeneID")

merged_TPM.filt <-
  as.matrix(Ec_abundance)
```

```{r}
log2_TPM <- log2(merged_TPM.filt + 1)

log2_TPM_renamed <- log2_TPM
base::colnames(log2_TPM_renamed) <- sample_info_ec_proc$LibLab
```

```{r}
ncol_Female <- grep( "F_" , colnames( log2_TPM_renamed ) )
ncol_Male <- grep( "M_" , colnames( log2_TPM_renamed ) )
```

#### Pearson correlation

```{r fig.height = 1.5, fig.width = 5}
M = cor(x = log2_TPM_renamed[,ncol_Female],
        y = log2_TPM_renamed[,ncol_Male],
        method = "pearson")

M_melt <- reshape2::melt(M) %>%
        dplyr::group_by(Var1, Var2) %>% 
        dplyr::summarise(median_corr = median(value)) %>%
        dplyr::filter(stringr::str_remove(Var2, "M_") == stringr::str_remove(Var1, "F_"))

M_melt_processed <- M_melt %>% mutate(Stage = stringr::str_remove(Var2, "M_"))
M_melt_processed$Stage <- factor(M_melt_processed$Stage, levels = unique(M_melt_processed$Stage))

ggplot2::ggplot(M_melt_processed, aes(x = Stage, y = 1, fill = median_corr, label = round(median_corr,2))) +
        ggplot2::geom_tile() +
        ggplot2::geom_text() +
        ggplot2::theme_minimal() +
        ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
                       legend.position = "",
                       axis.text.y=element_blank()) +
        ggplot2::scale_fill_gradient(low = "white", high = "#DC0000FF") +
        ggplot2::labs(
                x = "Ectocarpus stages",
                y = "Pearson")
```


#### Spearman correlation

```{r fig.height = 1.5, fig.width = 5}
M = cor(x = log2_TPM_renamed[,ncol_Female],
        y = log2_TPM_renamed[,ncol_Male],
        method = "spearman")

M_melt <- reshape2::melt(M) %>%
        dplyr::group_by(Var1, Var2) %>% 
        dplyr::summarise(median_corr = median(value)) %>%
        dplyr::filter(stringr::str_remove(Var2, "M_") == stringr::str_remove(Var1, "F_"))

M_melt_processed <- M_melt %>% mutate(Stage = stringr::str_remove(Var2, "M_"))
M_melt_processed$Stage <- factor(M_melt_processed$Stage, levels = unique(M_melt_processed$Stage))

ggplot2::ggplot(M_melt_processed, aes(x = Stage, y = 1, fill = median_corr, label = round(median_corr,2))) +
        ggplot2::geom_tile() +
        ggplot2::geom_text() +
        ggplot2::theme_minimal() +
        ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
                       legend.position = "",
                       axis.text.y=element_blank()) +
        ggplot2::scale_fill_gradient(low = "white", high = "#DC0000FF") +
        ggplot2::labs(
                x = "Ectocarpus stages",
                y = "Spearman")
```


#### JSD metric

```{r}
mat_normalized <- apply(log2_TPM_renamed, 2, function(x) x/sum(x)) %>% as.data.frame()
DM <- philentropy::distance(t(mat_normalized), use.row.names = TRUE, method = "jensen-shannon")
DM2 <- sqrt(DM)
```

```{r fig.height = 1.5, fig.width = 5}
M_melt <- reshape2::melt(DM2) %>%
        dplyr::group_by(Var1, Var2) %>% 
        dplyr::summarise(median_corr = median(value)) %>%
        dplyr::filter(stringr::str_remove(Var2, "M_") == stringr::str_remove(Var1, "F_"))

M_melt_processed <- M_melt %>% mutate(Stage = stringr::str_remove(Var2, "M_"))
M_melt_processed$Stage <- factor(M_melt_processed$Stage, levels = unique(M_melt_processed$Stage))

ggplot2::ggplot(M_melt_processed, aes(x = Stage, y = 1, fill = median_corr, label = round(median_corr,3))) +
        ggplot2::geom_tile() +
        ggplot2::geom_text() +
        ggplot2::theme_minimal() +
        ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
                       legend.position = "",
                       axis.text.y=element_blank()) +
        ggplot2::scale_fill_gradient(low = "white", high = "#3C5488FF") +
        ggplot2::labs(
                x = "Ectocarpus stages",
                y = "JSD")
```

#### Manhattan distance

```{r}
DM <- philentropy::distance(t(log2_TPM_renamed), use.row.names = TRUE, method = "manhattan")
```

```{r fig.height = 1.5, fig.width = 5}
M_melt <- reshape2::melt(DM) %>%
        dplyr::group_by(Var1, Var2) %>% 
        dplyr::summarise(median_corr = median(value)) %>%
        dplyr::filter(stringr::str_remove(Var2, "M_") == stringr::str_remove(Var1, "F_"))

M_melt_processed <- M_melt %>% mutate(Stage = stringr::str_remove(Var2, "M_"))
M_melt_processed$Stage <- factor(M_melt_processed$Stage, levels = unique(M_melt_processed$Stage))

ggplot2::ggplot(M_melt_processed, aes(x = Stage, y = 1, fill = median_corr, label = round(median_corr,0))) +
        ggplot2::geom_tile() +
        ggplot2::geom_text() +
        ggplot2::theme_minimal() +
        ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
                       legend.position = "",
                       axis.text.y=element_blank()) +
        ggplot2::scale_fill_gradient(low = "white", high = "#3C5488FF") +
        ggplot2::labs(
                x = "Ectocarpus stages",
                y = "Manhattan")
```

## denoised data

```{r}
Ec_abundance.denoised <-
        readr::read_csv(file = "data/Ec_abundance_denoised.csv") %>%
        dplyr::select(1, sample_info_ec_proc$SampleName) %>%
        tibble::column_to_rownames(var = "GeneID")

merged_TPM.filt <-
  as.matrix(Ec_abundance.denoised)
```

```{r}
log2_TPM <- log2(merged_TPM.filt + 1)

log2_TPM_renamed <- log2_TPM
base::colnames(log2_TPM_renamed) <- sample_info_ec_proc$LibLab
```

```{r}
ncol_Female <- grep( "F_" , colnames( log2_TPM_renamed ) )
ncol_Male <- grep( "M_" , colnames( log2_TPM_renamed ) )
```

#### Pearson correlation

```{r fig.height = 5, fig.width = 5}
M = cor(x = log2_TPM_renamed[,ncol_Female],
        y = log2_TPM_renamed[,ncol_Male],
        method = "pearson")

reshape2::melt(M) %>%
        dplyr::group_by(Var1, Var2) %>% 
        dplyr::summarise(median_corr = median(value)) %>%
        ggplot2::ggplot(aes(x = Var1, y = Var2, fill = median_corr, label = round(median_corr,2))) +
        ggplot2::geom_tile() +
        ggplot2::geom_text() +
        ggplot2::theme_minimal() +
        ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
                       legend.position = "") +
        ggplot2::scale_fill_gradient(low = "white", high = "#DC0000FF") +
        ggplot2::labs(
                x = "Female stages",
                y = "Male stages",
                title = "Pearson correlation")

```

```{r fig.height = 1.5, fig.width = 5}
M_melt <- reshape2::melt(M) %>%
        dplyr::group_by(Var1, Var2) %>% 
        dplyr::summarise(median_corr = median(value)) %>%
        dplyr::filter(stringr::str_remove(Var2, "M_") == stringr::str_remove(Var1, "F_"))

M_melt_processed <- M_melt %>% mutate(Stage = stringr::str_remove(Var2, "M_"))
M_melt_processed$Stage <- factor(M_melt_processed$Stage, levels = unique(M_melt_processed$Stage))

ggplot2::ggplot(M_melt_processed, aes(x = Stage, y = 1, fill = median_corr, label = round(median_corr,2))) +
        ggplot2::geom_tile() +
        ggplot2::geom_text() +
        ggplot2::theme_minimal() +
        ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
                       legend.position = "",
                       axis.text.y=element_blank()) +
        ggplot2::scale_fill_gradient(low = "white", high = "#DC0000FF") +
        ggplot2::labs(
                x = "Ectocarpus stages",
                y = "Pearson",
                fill = "Correlation")
```

#### Spearman correlation

```{r fig.height = 5, fig.width = 5}
M = cor(x = log2_TPM_renamed[,ncol_Female],
        y = log2_TPM_renamed[,ncol_Male],
        method = "spearman")

reshape2::melt(M) %>%
        dplyr::group_by(Var1, Var2) %>% 
        dplyr::summarise(median_corr = median(value)) %>%
        ggplot2::ggplot(aes(x = Var1, y = Var2, fill = median_corr, label = round(median_corr,2))) +
        ggplot2::geom_tile() +
        ggplot2::geom_text() +
        ggplot2::theme_minimal() +
        ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
                       legend.position = "") +
        ggplot2::scale_fill_gradient(low = "white", high = "#DC0000FF") +
        ggplot2::labs(
                x = "Female stages",
                y = "Male stages",
                title = "Spearman correlation")

```

```{r fig.height = 1.5, fig.width = 5}
M_melt <- reshape2::melt(M) %>%
        dplyr::group_by(Var1, Var2) %>% 
        dplyr::summarise(median_corr = median(value)) %>%
        dplyr::filter(stringr::str_remove(Var2, "M_") == stringr::str_remove(Var1, "F_"))

M_melt_processed <- M_melt %>% mutate(Stage = stringr::str_remove(Var2, "M_"))
M_melt_processed$Stage <- factor(M_melt_processed$Stage, levels = unique(M_melt_processed$Stage))

ggplot2::ggplot(M_melt_processed, aes(x = Stage, y = 1, fill = median_corr, label = round(median_corr,2))) +
        ggplot2::geom_tile() +
        ggplot2::geom_text() +
        ggplot2::theme_minimal() +
        ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
                       legend.position = "",
                       axis.text.y=element_blank()) +
        ggplot2::scale_fill_gradient(low = "white", high = "#DC0000FF") +
        ggplot2::labs(
                x = "Ectocarpus stages",
                y = "Spearman",
                fill = "Correlation")
```

#### JSD metric

```{r}
mat_normalized <- apply(log2_TPM_renamed, 2, function(x) x/sum(x)) %>% as.data.frame()
DM <- philentropy::distance(t(mat_normalized), use.row.names = TRUE, method = "jensen-shannon")
DM2 <- sqrt(DM)
```

```{r fig.height = 1.5, fig.width = 5}
M_melt <- reshape2::melt(DM2) %>%
        dplyr::group_by(Var1, Var2) %>% 
        dplyr::summarise(median_corr = median(value)) %>%
        dplyr::filter(stringr::str_remove(Var2, "M_") == stringr::str_remove(Var1, "F_"))

M_melt_processed <- M_melt %>% mutate(Stage = stringr::str_remove(Var2, "M_"))
M_melt_processed$Stage <- factor(M_melt_processed$Stage, levels = unique(M_melt_processed$Stage))

ggplot2::ggplot(M_melt_processed, aes(x = Stage, y = 1, fill = median_corr, label = round(median_corr,3))) +
        ggplot2::geom_tile() +
        ggplot2::geom_text() +
        ggplot2::theme_minimal() +
        ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
                       legend.position = "",
                       axis.text.y=element_blank()) +
        ggplot2::scale_fill_gradient(low = "white", high = "#3C5488FF") +
        ggplot2::labs(
                x = "Ectocarpus stages",
                y = "JSD")
```

#### Manhattan distance

```{r}
DM <- philentropy::distance(t(log2_TPM_renamed), use.row.names = TRUE, method = "manhattan")
```

```{r fig.height = 1.5, fig.width = 5}
M_melt <- reshape2::melt(DM) %>%
        dplyr::group_by(Var1, Var2) %>% 
        dplyr::summarise(median_corr = median(value)) %>%
        dplyr::filter(stringr::str_remove(Var2, "M_") == stringr::str_remove(Var1, "F_"))

M_melt_processed <- M_melt %>% mutate(Stage = stringr::str_remove(Var2, "M_"))
M_melt_processed$Stage <- factor(M_melt_processed$Stage, levels = unique(M_melt_processed$Stage))

ggplot2::ggplot(M_melt_processed, aes(x = Stage, y = 1, fill = median_corr, label = round(median_corr,0))) +
        ggplot2::geom_tile() +
        ggplot2::geom_text() +
        ggplot2::theme_minimal() +
        ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
                       legend.position = "",
                       axis.text.y=element_blank()) +
        ggplot2::scale_fill_gradient(low = "white", high = "#3C5488FF") +
        ggplot2::labs(
                x = "Ectocarpus stages",
                y = "Manhattan")
```

```{r}
devtools::session_info()
```