---
title: "4.1 OG DEG"
output:
  html_document:
    fig_width: 6
    fig_height: 4
    df_print: paged
editor_options: 
  chunk_output_type: inline
---

# Orthogroup DEGs

This is an orthologous analysis to the results presented in `4.0 OG visualisation`. Here we want to know how many genes are differentially expressed between OGs. This is admittedly a strang approach but I think none of the previous assumptions for gene-level data are violated by OG-level data.

```{r}
library(tidyverse)
library(DESeq2)
```

All the counts are based on Orthogroups

```{r}
Fs_counts <-
  readr::read_csv(file = "data/Fs_OG_counts.csv")
Fd_counts <-
  readr::read_csv(file = "data/Fd_OG_counts.csv")
Ec_counts <-
  readr::read_csv(file = "data/Ec_OG_counts.csv")
```

```{r}
sample_info_fd <-
  readr::read_csv("data/sample_info_fd.csv")
sample_info_fs <-
  readr::read_csv("data/sample_info_fs.csv")
sample_info_ec <-
  readr::read_csv("data/sample_info_ec.csv")

sample_info_fd_embryo <-
  sample_info_fd %>%
  dplyr::filter(Stage %in% c("E1", "E2", "E3", "E4", "E5")) %>%
  dplyr::mutate(LibLab = str_c(Species, "_", Stage, "_", Replicate))
sample_info_fs_embryo <-
  sample_info_fs %>%
  dplyr::filter(Stage %in% c("24H", "48H", "1w", "3w", "4w")) %>%
  dplyr::mutate(Replicate = case_when(
    LibName == "Fs_X_zygotes_48h_1" ~ 4,
    LibName == "Fs_X_zygotes_48h_2" ~ 5,
    LibName == "Fs_X_zygotes_48h_3" ~ 6,
    TRUE ~ Replicate
  )) %>%
  mutate(LibLab = str_c(Species, "_", Stage, "_", Replicate))
```

## Similarity between distributions (OG vs gene)

```{r}
Fs_counts_genewise <-
  readr::read_csv(file = "data/Fs_counts.csv")
```

Compare gene-wise to OG-wise count data.

```{r}
dist_pivoted <-
  tibble::tibble(feature_name = Fs_counts$OG, count = Fs_counts$Fs_zygotes_1w_1, feature_type = "OG_wise") %>%
  dplyr::add_row(tibble::tibble(feature_name = Fs_counts_genewise$GeneID, count = Fs_counts_genewise$Fs_zygotes_1w_1, feature_type = "gene_wise"))

dist_plot1 <- dist_pivoted %>%
  ggplot2::ggplot(aes(x=count, colour=feature_type, fill=feature_type)) + 
  ggplot2::geom_density(alpha = 0.5, colour = "black") +
  ggplot2::labs(
    # title = "Distribution of OG-wise \nand gene-wise count data",
    subtitle = "Sample: Fs_zygotes_1w_1",
    x = "count (log-scaled)"
  ) +
  ggplot2::xlim(1, 1e+05) +
  ggplot2::scale_x_log10()

dist_pivoted <-
  tibble::tibble(feature_name = Fs_counts$OG, count = Fs_counts$Fs_zygotes_4w_1, feature_type = "OG_wise") %>%
  dplyr::add_row(tibble::tibble(feature_name = Fs_counts_genewise$GeneID, count = Fs_counts_genewise$Fs_zygotes_4w_1, feature_type = "gene_wise"))

dist_plot2 <- dist_pivoted %>%
  ggplot2::ggplot(aes(x=count, colour=feature_type, fill=feature_type)) + 
  ggplot2::geom_density(alpha = 0.5, colour = "black") +
  ggplot2::labs(
    # title = "Distribution of OG-wise \nand gene-wise count data",
    subtitle = "Sample: Fs_zygotes_4w_1",
    x = "count (log-scaled)"
  ) +
  ggplot2::xlim(1, 1e+05) +
  ggplot2::scale_x_log10()

dist_pivoted <-
  tibble::tibble(feature_name = Fs_counts$OG, count = Fs_counts$Fs_F_gametes_1, feature_type = "OG_wise") %>%
  dplyr::add_row(tibble::tibble(feature_name = Fs_counts_genewise$GeneID, count = Fs_counts_genewise$Fs_F_gametes_1, feature_type = "gene_wise"))

dist_plot3 <- dist_pivoted %>%
  ggplot2::ggplot(aes(x=count, colour=feature_type, fill=feature_type)) + 
  ggplot2::geom_density(alpha = 0.5, colour = "black") +
  ggplot2::labs(
    # title = "Distribution of OG-wise \nand gene-wise count data",
    subtitle = "Sample: Fs_F_gametes_1",
    x = "count (log-scaled)"
  ) +
  ggplot2::xlim(1, 1e+05) +
  ggplot2::scale_x_log10()

dist_pivoted <-
  tibble::tibble(feature_name = Fs_counts$OG, count = Fs_counts$Fs_Fwild_matSP_holdfast_2, feature_type = "OG_wise") %>%
  dplyr::add_row(tibble::tibble(feature_name = Fs_counts_genewise$GeneID, count = Fs_counts_genewise$Fs_Fwild_matSP_holdfast_2, feature_type = "gene_wise"))

dist_plot4 <- dist_pivoted %>%
  ggplot2::ggplot(aes(x=count, colour=feature_type, fill=feature_type)) + 
  ggplot2::geom_density(alpha = 0.5, colour = "black") +
  ggplot2::labs(
    # title = "Distribution of OG-wise \nand gene-wise count data",
    subtitle = "Sample: Fs_Fwild_matSP_holdfast_2",
    x = "count (log-scaled)"
  ) +
  ggplot2::xlim(1, 1e+05) +
  ggplot2::scale_x_log10()

ggpubr::ggarrange(dist_plot1, dist_plot2, dist_plot3, dist_plot4, common.legend = TRUE) %>% ggpubr::annotate_figure(top = ggpubr::text_grob("Distribution of OG-wise and gene-wise count data", size = 14))
```

We see here that the distributions are similar based on the log2-transformed data. I assume that this is the case in different samples.

## OG DESeq2

```{r}
compare_groups <- function(dds, contrast_groups, altHypothesis = "greaterAbs", padj_threshold = 0.1, lfcThreshold = 1){
  INPUT_GROUPS <- unique(contrast_groups)
  OUT <- tibble()
  for (i in INPUT_GROUPS) {
    for (j in INPUT_GROUPS[INPUT_GROUPS != i]) {
      DESeq2_res <- results(dds, contrast=c("group", i, j), altHypothesis = altHypothesis, lfcThreshold = lfcThreshold)
      DESeq2_res <- DESeq2_res %>%
        as_tibble(rownames = "GeneID") %>%
        drop_na() %>%
        group_by(SIG = padj <= 0.1)
      
      SIG_res <- tibble(group1 = paste(i), group2 = paste(j), significant = sum(DESeq2_res$SIG == T), nonsignificant = sum(DESeq2_res$SIG == F))
      # SIG_res <- tibble(group1 = paste(i), group2 = paste(j))
      
      OUT <- rbind(OUT, SIG_res)
    }
  }
  return(OUT)
}
```

```{r}
selectGenes <- function(counts, min.count=2){
  keep <-
    counts[rowMeans(counts)>min.count, ]
  return(keep)
}
```

Here are the potential models we could use to compare.

```{r}
# simple.model <- as.formula(~ Stage)
# additive.model <- as.formula(~ Species + Stage)
# interactive.model <- as.formula(~ Species:Stage)
complex.model <- as.formula(~ Species + Species:Stage)
```

```{r}
combined_metatable <- 
  rbind(sample_info_fs_embryo, sample_info_fd_embryo) %>%
  relocate(LibName) %>%
  dplyr::mutate(Stage = stage_tissue)
merged_count.pre <- 
  merge(Fs_counts, Fd_counts, by = "OG") %>%
  dplyr::select(1, combined_metatable$LibName)
merged_count <- 
  data.frame(row.names = merged_count.pre[,1], merged_count.pre[,-1], check.names = FALSE)
merged_count.filt <-
  as.matrix(merged_count) %>%
  selectGenes(min.count=5)
combined_metatable$Stage <- 
  factor(combined_metatable$Stage, 
         levels = unique(combined_metatable$Stage))
```


```{r}
ddsObj.raw_combined <- 
  DESeqDataSetFromMatrix(
    round(merged_count.filt,0),
    colData = combined_metatable,
    design = complex.model)

ddsObj.raw_combined$group <- factor(paste0(ddsObj.raw_combined$Species, ddsObj.raw_combined$Stage))
design(ddsObj.raw_combined) <- ~ group
dds <- DESeq(ddsObj.raw_combined)
resultsNames(dds)

```

```{r}
DEGs <- compare_groups(dds, ddsObj.raw_combined$group)
nonDEGs <- compare_groups(dds, ddsObj.raw_combined$group, altHypothesis = "lessAbs", lfcThreshold = 1)
```

```{r}
DEGs_proc <- DEGs %>% 
  mutate(prop_sig = significant/(significant + nonsignificant)) %>%
  # select(-3,-4) %>%
  dplyr::filter(str_starts(group2, "Fd")) %>%
  dplyr::filter(str_starts(group1, "Fs")) %>%
  mutate(group2 = str_replace(group2, "24H", "E1")) %>% 
  mutate(group2 = str_replace(group2, "48H", "E2")) %>%
  mutate(group2 = str_replace(group2, "1w", "E3")) %>%
  mutate(group2 = str_replace(group2, "3w", "E4")) %>%
  mutate(group2 = str_replace(group2, "4w", "E5"))

nonDEGs_proc <- nonDEGs %>% 
  mutate(prop_sig = significant/(significant + nonsignificant)) %>%
  # select(-3,-4) %>%
  dplyr::filter(str_starts(group2, "Fd")) %>%
  dplyr::filter(str_starts(group1, "Fs")) %>%
  mutate(group2 = str_replace(group2, "24H", "E1")) %>% 
  mutate(group2 = str_replace(group2, "48H", "E2")) %>%
  mutate(group2 = str_replace(group2, "1w", "E3")) %>%
  mutate(group2 = str_replace(group2, "3w", "E4")) %>%
  mutate(group2 = str_replace(group2, "4w", "E5"))

my_palette <- colorRampPalette(c("white", "#E64B35FF"))(n = 100)

DEGs_proc$group1 <- factor(DEGs_proc$group1, levels = unique(DEGs_proc$group1))
DEGs_proc$group2 <- factor(DEGs_proc$group2, levels = unique(DEGs_proc$group2))
nonDEGs_proc$group1 <- factor(nonDEGs_proc$group1, levels = unique(nonDEGs_proc$group1))
nonDEGs_proc$group2 <- factor(nonDEGs_proc$group2, levels = unique(nonDEGs_proc$group2))
```

```{r, results='hide', warning=FALSE, message=FALSE, fig.height = 3, fig.width = 6}
ggplot(DEGs_proc, aes(x = group1, y = group2, fill = prop_sig*100)) +
  geom_tile(color = "white") +
  scale_fill_gradientn(colors = my_palette) +
  geom_text(aes(label = paste0(significant, "/", (significant + nonsignificant))), color = "black") +
  labs(title = "DEG",
       x = "Fucus serratus stages",
       y = "Fucus distichus stages",
       fill = "% significant") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5)) + 
  theme_minimal()

ggplot(nonDEGs_proc, aes(x = group1, y = group2, fill = prop_sig*100)) +
  geom_tile(color = "white") +
  scale_fill_gradientn(colors = my_palette) +
  geom_text(aes(label = paste0(significant, "/", (significant + nonsignificant))), color = "black") +
  labs(title = "nonDEG",
       x = "Fucus serratus stages",
       y = "Fucus distichus stages",
       fill = "% significant") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5)) + 
  theme_minimal()
```


```{r, results='hide', warning=FALSE, message=FALSE, fig.height = 3, fig.width = 5}
ggplot(DEGs_proc, aes(x = group1, y = group2, fill = prop_sig)) +
  geom_tile(color = "white") +
  scale_fill_gradientn(colors = my_palette) +
  geom_text(aes(label = round(prop_sig,4)), color = "black") +
  labs(title = "DEG",
       x = "Fucus serratus stages",
       y = "Fucus distichus stages",
       fill = "prop. significant") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5)) + 
  theme_minimal()

ggplot(nonDEGs_proc, aes(x = group1, y = group2, fill = prop_sig)) +
  geom_tile(color = "white") +
  scale_fill_gradientn(colors = my_palette) +
  geom_text(aes(label = round(prop_sig,4)), color = "black") +
  labs(title = "nonDEG",
       x = "Fucus serratus stages",
       y = "Fucus distichus stages",
       fill = "prop. significant") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5)) + 
  theme_minimal()
```

Was wir einfach sehen können ist dass das Muster mit der differenziellen Genexpression das widerspiegelt, was wir mit Distanzmetriken sehen können.

```{r}
devtools::session_info()
```